### PR TITLE
ENH: add support for in-place array modification to apply_along_axis

### LINKED ISSUE #13192: better support for in-place modification in apply_along_axis
In #13149 it was observed that the call to `apply_along_axis` within `np.pad` assumes in-place modification of the array. In this case, `apply_along_axis` itself will still allocate a new output array, which is inefficient. It would be nice to have some sort of `inplace` argument for `apply_along_axis` , to indicate when no output array needs to be created.

I will create a PR with a minimal implementation to have something concrete to discuss. I think the main tricky point is how to best set up the API in a way that doesn't potentially break existing code.

### PR DESCRIPTION
**Pull request by grlee77**
This PR implements an option to indicate to `apply_along_axis` that the function modifies `arr` in-place.

closes #13192 

I am not sure about how best to handle the argument or what to name it, but for now I added an optional kwarg called `inplace`.

I didn't want to add any new positional arguments as that would almost certainly break existing user code. The `inplace` kwarg will also potentially break user code if a user-supplied `func1d` also expected a kwarg named `inplace`, but that is probably much less likely.

Another option would be to "hide" the kwarg by naming it something like `_inplace` if we only wanted to only support it for internal use such as in `np.pad` (I did not find other locations in NumPy itself where in-place modification was being used).

Some test cases that indicate the expected behavior were added.

### CODE REVIEW

**Comment by grlee77** [comment]
I added a commit with an alternative solution that I think is perhaps cleaner. In that case, no new argument/option is added. Instead, in-place operation is assumed when `func1d` returns `None`.

**Comment by grlee77** [comment]
If/when we decide on the approach I can clean up the commit history (I had made changes to an outdated copy of arraypad.py in one of the earlier commits)

File: numpy/lib/shape_base.py [hunk 1]
```diff
@@ -376,7 +377,15 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
         ind0 = next(inds)
     except StopIteration:
         raise ValueError('Cannot apply_along_axis when any iteration dimensions are 0')
-    res = asanyarray(func1d(inarr_view[ind0], *args, **kwargs))
+
+    res = func1d(inarr_view[ind0], *args, **kwargs)
+    if res is None:
+        # no return value: assume arr is being modified in-place
```
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: Not a safe assumption, `func1d` could return a mixture of `None`s and other objects

Comment (grlee77) [reply numpy/lib/shape_base.py]: I'm not entirely sure I follow you. You mean in that checking only the first call to func1d may not be representative of the return value for all locations?

If you think it is preferable, I can revert to a keyword-based approach as implemented in 7d910da

Comment (eric-wieser) [reply numpy/lib/shape_base.py]: It's not clear to me that this function makes sense at all. Why are we working a glorified loop with a callback and not an iterator?

**Comment by grlee77** [comment]
I am not sure that this addition to `apply_along_axis` would be used often.

See #13250 for an alternative solution that just modifies `np.pad` to avoid the overhead of `apply_along_axis`. I could not find any other instances in the code base where in-place operation was assumed for `apply_along_axis`.

**Comment by grlee77** [comment]
closing in favor of #13250

**Pull request closed without merging**